qa_handler.hpp

```diff
@@ -17,61 +18,72 @@
 using namespace taginfo_validate;
 
 struct qa_handler : osmium::handler::Handler {
-  qa_handler(const taginfo_parser &taginfo)
+  qa_handler(const taginfo_parser &taginfo, std::unordered_map<std::string, uint32_t> &string_catalogue)
       : tags_on_areas(taginfo.tags_on_areas()), tags_on_nodes(taginfo.tags_on_nodes()),
         tags_on_ways(taginfo.tags_on_ways()), tags_on_relations(taginfo.tags_on_relations()),
-        tags_on_any_object(taginfo.tags_on_any_object()){};
+        tags_on_any_object(taginfo.tags_on_any_object()), ST(string_catalogue){};
 
   using tag_range = std::pair<taginfo_parser::tag_iter, taginfo_parser::tag_iter>;
   struct keyCompare {
     bool operator()(const tag &left, const tag &right) { return left.key < right.key; }
   };
 
-  // if pbf tag is not found in taginfo list, store tag to list of unrecognized tags
+  // Given an object type, a range of tags of that object type and a pbf tag, check
+  // if the pbf tag is found in the range of tags, or store the pbf tag to list of unrecognized tags
   void verifyAndStoreTags(const object::type &objectType, const tag_range &tagRange, const osmium::Tag &thePbfTag) {
+    // first store in tag dictionary
+    if (ST.find(thePbfTag.key()) == ST.end()) {
+      ST[thePbfTag.key()] = ST.size();
+    }
+    if (ST.find(thePbfTag.value()) == ST.end()) {
+      ST[thePbfTag.value()] = ST.size();
+    }
     // Look for the given pbf tag's key in the range of acceptable keys
-    auto matching_key_tags = std::equal_range(tagRange.first, tagRange.second, tag{thePbfTag.key(), {}, objectType}, keyCompare());
+    auto matching_key_tags =
+        std::equal_range(tagRange.first, tagRange.second, tag{ST[thePbfTag.key()], 0, objectType}, keyCompare());
     auto matching_key_tags_anyObject = std::equal_range(tags_on_any_object.first, tags_on_any_object.second,
-                                                        tag{thePbfTag.key(), {}, object::type::all}, keyCompare());
+                                                        tag{ST[thePbfTag.key()], 0, object::type::all}, keyCompare());
```

Comment:
Use `ST.at(kv)` instead. `ST[kv]` checks if `kv` is in the map, if not it creates a default initialized item and returns a reference to it.